search/models.py

```diff
@@ -0,0 +1,120 @@
+from django.db import models
+from django.db.backends.signals import connection_created
+from django.db.models.signals import post_save, pre_delete
+from django.dispatch import receiver
+from django.utils.module_loading import import_string
+
+from .utils import rank
+
+
+class Match(models.Lookup):
+    lookup_name = 'match'
+
+    def as_sql(self, qn, connection):
+        lhs, lhs_params = self.process_lhs(qn, connection)
+        rhs, rhs_params = self.process_rhs(qn, connection)
+        params = lhs_params + rhs_params
+        return '%s MATCH %s' % (lhs, rhs), params
+
+
+class SearchField(models.Field):
+    pass
+SearchField.register_lookup(Match)
+
+
+class SearchQuerySet(models.QuerySet):
+    def order_by_relevancy(self):
+        extra = {'relevancy': 'rank(matchinfo(idx))'}
+        return self.extra(select=extra).order_by('-relevancy')
+
+
+class Search(models.Model):
+    """Model that handle the search."""
+    rowid = models.IntegerField(primary_key=True)
+    module = models.CharField(max_length=64)
+    model = models.CharField(max_length=64)
+    model_id = models.IntegerField()
+    public = models.BooleanField(default=True)
+    text = SearchField()
+
+    objects = SearchQuerySet.as_manager()
+
+    class Meta:
+        db_table = 'idx'
+        managed = False
+
+    @classmethod
+    def ids(cls, **kwargs):
+        qs = Search.objects.filter(**kwargs).order_by_relevancy()
+        return qs.values_list('model_id', flat=True)
+
+    @classmethod
+    def search(cls, **kwargs):
+        qs = Search.objects.filter(**kwargs).order_by_relevancy()
+        for row in qs:
+            key = '.'.join([row.module, row.model])
+            if not key in _SEARCHABLE:
+                _SEARCHABLE[key] = import_string(key)
+            yield _SEARCHABLE[key].objects.get(pk=row.model_id)
+_SEARCHABLE = {}
```

Comment:
Humm, but if I want this to be dynamic, I need to loop over the result to know which models are available and need to be imported.

Another option, is to have them hardcoded. DRY alert, but maybe less ugly than the on the fly caching, and less complex/overkill than a metaclass plus a registry.